doc/developer/design/20230306_persist_mfp_pushdown.md

```diff
@@ -0,0 +1,272 @@
+- Third, we can adjust truncation of stats over varlen columns. For example, a bound of `a<huge string>..=y<huge string>` can likely be shortened to `a..=z`. This does make the bounds less tight, which increases false positives.
+- Finally, we can trim stats for `Record`, `List`, or `Jsonb` elements or even for entire columns. Note that if we do this, we almost certainly want some way for the customer to be able to hint that a certain column is critical and should never have trimmed stats (possibly even something as simple as a LaunchDarkly flag).
+
+
+Note that it is expected that the common case will be that stats do not need trimming. This is more a safety measure.
+
+## Filtering parts
+[filtering parts]: #filtering-parts
+
+Thanks to our part statistics, we know the range of possible values any particular column may have.
+To decide whether we can filter a part or not, though, we need to know the range of possible values of _the output of the MFP's filter:_ if we can prove that the filter will never return `true` or an error value, it's safe to skip the entire part.
+
+Using the range of possible input values to compute the range of possible output values is a well-known static analysis problem, and it's commonly solved via [abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation). In practice, this involves:
+- Choosing an "abstract set" to approximate your data type. For part filters, we can use roughly:
+    ```rust
+    /// Approximating a range of possible non-null datums.
+    enum Values<'a> {
+      Empty, // No non-null datums.
+      Within(Datum<'a>, Datum<'a>), // An inclusive range.
+      All, // Includes all datums.
+    }
+
+    /// Approximating a set of possible `Result<Datum, Err>`;
+    /// ie. the result of `MirScalarExpr::eval`.
+    pub struct ResultSpec<'a> {
+      nullable: bool, // Could this result be null?
+      fallible: bool, // Could this result be an error?
+      values: Values<'a>, // What range of non-null values may be included?
+    }
+    ```
+    This is meant to be compact enough to compute efficiently, but still precise enough to produce interesting results for many expressions.
+- Writing an "interpreter" - for every possible expression, we need to be able to evaluate that expression as an instance of our `ResultSpec`. (As compared to the normal concrete interpreter, where the expression is evaluated as an ordinary `Result`.)
+  - `Literal`s are interpreted directly: eg. `Datum::Null` becomes `ResultSpec { nullable: true, ..}`.
+  - For `Column`s, we know the set of possible values from the statistics. For `UnmaterializableFunc`s like `mz_now()`, we can also provide a range of possible values up front.
```

Comment:
@mjibson is proposing some changes to our optimization pipeline that will move it largely off-thread and only determine the `mz_now()` value after that and prior to `mir_to_lir` rendering.